mlflow/pipelines/steps/train.py

```diff
@@ -124,135 +124,181 @@ def _run(self, output_directory):
+                except ModuleNotFoundError:
                     raise MlflowException(
-                        f"Parameter {param_name} must contain either a list of 'values' or a "
-                        f"'distribution' following hyperopt parameter expressions",
-                        error_code=INVALID_PARAMETER_VALUE,
+                        "Hyperopt not installed and is required if tuning is enabled",
+                        error_code=BAD_REQUEST,
                     )
-            # minimize
-            algorithm = tuning_params["algorithm"]  # pylint: disable=unused-variable
-            max_trials = tuning_params["max_trials"]  # pylint: disable=unused-variable
-            best = fmin(  # pylint: disable=unused-variable
-                objective, search_space, max_evals=max_trials
-            )
 
-        else:
-            estimator = estimator_fn()
-            mlflow.autolog(log_models=False)
-
-            with mlflow.start_run(tags=tags) as run:
-                estimator.fit(X_train, y_train)
-
-                if hasattr(estimator, "best_score_"):
-                    mlflow.log_metric("best_cv_score", estimator.best_score_)
-                if hasattr(estimator, "best_params_"):
-                    mlflow.log_params(estimator.best_params_)
-
-                code_paths = [os.path.join(self.pipeline_root, "steps")]
-                estimator_schema = infer_signature(X_train, estimator.predict(X_train.copy()))
-                logged_estimator = mlflow.sklearn.log_model(
-                    estimator,
-                    f"{self.name}/estimator",
-                    signature=estimator_schema,
-                    code_paths=code_paths,
+                # wrap training in objective fn
+                def objective(hyperparameter_args):
+                    # log as a child run
+                    with mlflow.start_run(tags=tags, nested=True):
+                        # create unfitted estimator from yaml
+                        estimator = estimator_fn(
+                            dict(estimator_hardcoded_params, **hyperparameter_args)
+                        )
+                        sample_fraction = (
+                            float(tuning_params["sample_fraction"])
+                            if "sample_fraction" in tuning_params
+                            else 1.0
+                        )
+                        X_train_sampled = X_train.sample(frac=sample_fraction, random_state=42)
+                        y_train_sampled = y_train.sample(frac=sample_fraction, random_state=42)
+                        # fit estimator to training
+                        estimator.fit(X_train_sampled, y_train_sampled)
+                        if hasattr(estimator, "best_score_P"):
+                            mlflow.log_metric("best_cv_score", estimator.best_score_)
+                        if hasattr(estimator, "best_params_"):
+                            mlflow.log_params(estimator.best_params_)
```

Comment:
Is it ever possible for these best params / best score to be invalid (e.g. `None`)? Shall we validate them?

Reply:
Ah, I see. This is duplicated from below. Can we factor it out?